app/server/endpoint/src/main/java/io/syndesis/server/endpoint/v1/handler/meta/AggregateMetadataHandler.java

```diff
@@ -47,34 +47,38 @@ public boolean canHandle(StepKind kind) {
 
     @Override
     public DynamicActionMetadata handle(DynamicActionMetadata metadata) {
-        try {
-            DataShape outputShape = metadata.outputShape();
-            DataShape inputShape = metadata.inputShape();
+        DataShape outputShape = metadata.outputShape();
+        DataShape inputShape = metadata.inputShape();
 
+        try {
             if (metadata.outputShape() != null) {
                 outputShape = adaptOutputShape(new DataShape.Builder()
-                                                        .createFrom(metadata.outputShape())
-                                                        .decompress()
-                                                        .build());
+                        .createFrom(metadata.outputShape())
+                        .decompress()
+                        .build());
             }
+        } catch (IOException e) {
+            LOG.warn("Unable to read input data shape on dynamic meta data inspection", e);
+            LOG.warn("Using original input shape as fallback");
+        }
 
+        try {
             if (metadata.inputShape() != null) {
                 inputShape = adaptInputShape(new DataShape.Builder()
                                                         .createFrom(metadata.inputShape())
                                                         .decompress()
                                                         .build());
             }
-
-            return new DynamicActionMetadata.Builder()
-                    .createFrom(metadata)
-                    .inputShape(inputShape)
-                    .outputShape(outputShape)
-                    .build();
         } catch (IOException e) {
             LOG.warn("Unable to read output data shape on dynamic meta data inspection", e);
+            LOG.warn("Using original output shape as fallback");
         }
 
-        return metadata;
+        return new DynamicActionMetadata.Builder()
+                .createFrom(metadata)
+                .inputShape(inputShape)
+                .outputShape(outputShape)
+                .build();
     }
 
     DataShape adaptInputShape(DataShape dataShape) throws IOException {
```

Comment:
I'd probable refactor this method into separate methods for cases where `specification` is given, `specification` is given and `kind` is `JSON_SCHEMA` and kind is `JSON_INSTANCE`.

Reply:
Yeah the method is too complicated, will do on master and push in a separate PR